GPflowOpt/objective.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2017 Joachim van der Herten
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import numpy as np
+from functools import wraps
+from GPflow import model
+
+
+def batch_apply(fun):
+    """
+    Decorator which applies a function along the first dimension of a given ndarray as argument (the batch dimension)
+    the most common use case is to convert a function designed to operate on a single input vector, and
+    to compute its response (and possibly gradient) for each row of a matrix.
+
+    :param fun: function accepting an input vector of dimensionality d and returns a vector of dimensionality p (the
+       output dimensionality) and (optionally) a gradient of size d x p (or d if p == 1)
+    :return: a function wrapper which calls fun on each row of a given n x d matrix. Here n represents the batch
+       dimension. the wrapper returns n x p and optionally a n x d x p matrix (or n x d if p == 1)
+    """
+    @wraps(fun)
+    def batch_wrapper(X):
+        responses = (fun(x) for x in np.atleast_2d(X))
+        sep = tuple(zip(*(r if isinstance(r, tuple) else (r,) for r in responses)))
+        f = np.vstack(sep[0])
+        if len(sep) == 1:
+            return f
+
+        # for each point, the gradient is either (d,) or (d, p) shaped.
+        g_stacked = np.stack((r for r in sep[1]), axis=0)  # n x d or n x d x p
+        # Get rid of last dim = 1 in case p = 1
+        g = np.squeeze(g_stacked, axis=2) if len(g_stacked.shape) == 3 and g_stacked.shape[2] == 1 else g_stacked
```

Comment:
Currently we don't have any gradient-based optimizers for multiple objectives, so the n x p x d case doesn't occur currently. Instead, right now n x d matrices are returned.